esphome/components/ags10/ags10.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include "esphome/core/automation.h"
+#include "esphome/core/component.h"
+#include "esphome/components/sensor/sensor.h"
+#include "esphome/components/i2c/i2c.h"
+
+namespace esphome {
+namespace ags10 {
+
+class AGS10Component : public PollingComponent, public i2c::I2CDevice {
+ public:
+  /**
+   * Sets TVOC sensor.
+   */
+  void set_tvoc(sensor::Sensor *tvoc) { tvoc_ = tvoc; }
+
+  /**
+   * Sets version info sensor.
+   */
+  void set_version(sensor::Sensor *version) { version_ = version; }
+
+  /**
+   * Sets resistance info sensor.
+   */
+  void set_resistance(sensor::Sensor *resistance) { resistance_ = resistance; }
+
+  void setup() override;
+
+  void update() override;
+
+  void dump_config() override;
+
+  float get_setup_priority() const override { return setup_priority::DATA; }
+
+  /**
+   * Modifies target address of AGS10.
+   *
+   * New address is saved and takes effect immediately even after power-off.
+   */
+  bool new_i2c_address(uint8_t newaddress);
+
+  /**
+   * Sets zero-point with factory defaults.
+   */
+  bool set_zero_point_with_factory_defaults();
+
+  /**
+   * Sets zero-point with current sensor resistance.
+   */
+  bool set_zero_point_with_current_resistance();
+
+  /**
+   * Sets zero-point with the value.
+   */
+  bool set_zero_point_with(uint16_t value);
+
+ protected:
+  /**
+   *  TVOC.
+   */
+  sensor::Sensor *tvoc_{nullptr};
+
+  /**
+   * Firmvare version.
+   */
+  sensor::Sensor *version_{nullptr};
+
+  /**
+   * Resistance.
+   */
+  sensor::Sensor *resistance_{nullptr};
+
+  /**
+   *  Last operation error code.
+   */
+  enum ErrorCode {
+    NONE = 0,
+    COMMUNICATION_FAILED,
+    CRC_CHECK_FAILED,
+    ILLEGAL_STATUS,
+    UNSUPPORTED_UNITS,
+  } error_code_{NONE};
+
+  /**
+   *  Reads and returns value of TVOC.
+   */
+  optional<uint32_t> read_tvoc_();
+
+  /**
+   *  Reads and returns a firmware version of AGS10.
+   */
+  optional<uint8_t> read_version_();
+
+  /**
+   * Reads and returns the resistance of AGS10.
+   */
+  optional<uint32_t> read_resistance_();
+
+  /**
+   * Read, checks and returns data from the sensor.
+   */
+  template<size_t N> optional<std::array<uint8_t, N>> read_and_check_(uint8_t a_register);
+
+  /**
+   * Calculates CRC8 value.
+   *
+   * CRC8 calculation, initial value: 0xFF, polynomial: 0x31 (x8+ x5+ x4+1)
+   *
+   * @param[in] dat the data buffer
+   * @param num number of bytes in the buffer
+   */
+  template<size_t N> uint8_t calc_crc8_(std::array<uint8_t, N> dat, uint8_t num);
+};
+
+template<typename... Ts> class AGS10NewI2cAddressAction : public Action<Ts...> {
+ public:
+  AGS10NewI2cAddressAction(AGS10Component *parent) : parent_(parent) {}
+  TEMPLATABLE_VALUE(int, new_address)
```

Comment:
I2C addresses are limites to a single byte, no need to reserve more memory that wont be used
```suggestion
  TEMPLATABLE_VALUE(uint8_t, new_address)
```

Reply:
Done